lib/charms/grafana_k8s/v0/grafana_source.py

```diff
@@ -0,0 +1,433 @@
+            self._stored.sources[rel_id]
+        )
+
+    def _update_sources(self, data: Dict, rel_id: int) -> None:
+        """Updates the event data bucket with new data"""
+        self._stored.sources[rel_id] = data
+        rel = self.framework.model.get_relation(self.name, rel_id)
+
+        logger.debug("Updating Grafana source data to {}".format(data))
+        rel.data[self.charm.app]["sources"] = json.dumps(data)
+
+
+class GrafanaSourceProvider(ProviderBase):
+    on = GrafanaSourceEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm: CharmBase, name: str, service: str, version=None) -> None:
+        """A Grafana based Monitoring service consumer
+
+        Args:
+            charm: a :class:`CharmBase` instance that manages this
+                instance of the Grafana source service.
+            name: string name of the relation that is provides the
+                Grafana source service.
+            service: string name of service provided. This is used by
+                :class:`GrafanaSourceProvider` to validate this service as
+                acceptable. Hence the string name must match one of the
+                acceptable service names in the :class:`GrafanaSourceProvider`s
+                `consumes` argument. Typically this string is just "grafana".
+            version: a string providing the semantic version of the Grafana
+                source being provided.
+
+        """
+        super().__init__(charm, name, service, version)
+        self.charm = charm
+        events = self.charm.on[name]
+
+        self._stored.set_default(sources=dict())  # available data sources
+        self._stored.set_default(sources_to_delete=set())
+
+        self.framework.observe(
+            events.relation_changed, self.on_grafana_source_relation_changed
+        )
+        self.framework.observe(
+            events.relation_broken, self.on_grafana_source_relation_broken
+        )
+
+    def on_grafana_source_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle relation changes in related consumers.
+
+        If there are changes in relations between Grafana source providers
+        and consumers, this event handler (if the unit is the leader) will
+        get data for an incoming grafana-source relation through a
+        :class:`GrafanaSourcesChanged` event, and make the relation data
+        is available in the app's datastore object. The Grafana charm can
+        then respond to the event to update its configuration
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        rel = event.relation
+        data = (
+            json.loads(event.relation.data[event.app].get("sources", {}))
+            if event.relation.data[event.app].get("sources", {})
+            else None
+        )
+        if not data:
+            logger.info("NO DATA")
+            return
+
+        # This is due to a double fire on relation_joined calling
```

Comment:
I am surprised. I was given to understand that if relation data is unchanged then Juju does not fire a relation changed event. Is this wrong ? If so could it be a Juju bug ? If it may be a bug, then a bug ticket link here would help in future maintenance of this lines.

Reply:
I may not be reading their repo right, but if I am, there is no DeepEquals against the map[string]interface{} parts before updating. I'll ask in discourse